File: echo/frontend/src/lib/passwordPolicy.ts
// Keep in sync with server/dembrane/password_policy.py and the
// auth_password_policy regex in directus/sync/collections/settings.json.

export const PASSWORD_MIN_LENGTH = 8;

export interface PasswordRules {
	lowercase: boolean;
	uppercase: boolean;
	number: boolean;
	symbol: boolean;
	minLength: boolean;
}

export type PasswordStrength = "weak" | "fair" | "strong";

export interface PasswordValidation {
	rules: PasswordRules;
	isValid: boolean;
	strength: PasswordStrength;
}

export function validatePassword(password: string): PasswordValidation {
	const rules: PasswordRules = {
		lowercase: /[a-z]/.test(password),
		minLength: password.length >= PASSWORD_MIN_LENGTH,
		number: /[0-9]/.test(password),
		symbol: /[^A-Za-z0-9]/.test(password),
		uppercase: /[A-Z]/.test(password),
	};

	const isValid = Object.values(rules).every(Boolean);
	const satisfied = Object.values(rules).filter(Boolean).length;

	let strength: PasswordStrength;
	if (!isValid) {
		strength = satisfied <= 2 ? "weak" : "fair";
	} else {
		strength = password.length >= 12 ? "strong" : "fair";
	}

	return { isValid, rules, strength };
}

File: echo/server/dembrane/password_policy.py
"""Password policy. Keep in sync with frontend/src/lib/passwordPolicy.ts and the
auth_password_policy regex in directus/sync/collections/settings.json."""
import re

PASSWORD_MIN_LENGTH = 8


def validate_password(password: str) -> list[str]:
    """Return a message per unmet rule. Empty list means the password is strong."""
    errors: list[str] = []
    if len(password) < PASSWORD_MIN_LENGTH:
        errors.append(f"Password must be at least {PASSWORD_MIN_LENGTH} characters")
    if not re.search(r"[a-z]", password):
        errors.append("Password must contain a lowercase letter")
    if not re.search(r"[A-Z]", password):
        errors.append("Password must contain an uppercase letter")
    if not re.search(r"[0-9]", password):
        errors.append("Password must contain a number")
    if not re.search(r"[^A-Za-z0-9]", password):
        errors.append("Password must contain a symbol")
    return errors


def is_strong_password(password: str) -> bool:
    return not validate_password(password)
